core/io/image.cpp

```diff
@@ -2719,6 +2718,33 @@ void Image::fill(const Color &c) {
 	}
 }
 
+void Image::fill_rect(const Rect2 &p_rect, const Color &p_color) {
+	ERR_FAIL_COND_MSG(!_can_modify(format), "Cannot fill rect in compressed or custom image formats.");
+
+	Rect2i r = Rect2i(0, 0, width, height).intersection(p_rect.abs());
+	if (r.has_no_area()) {
+		return;
+	}
+
+	uint8_t *dst_data_ptr = data.ptrw();
+
+	int pixel_size = get_format_pixel_size(format);
+
+	// Put first pixel with the format-aware API.
+	set_pixelv(r.position, p_color);
+	const uint8_t *rect_first_pixel_ptr = &dst_data_ptr[(r.position.y * width + r.position.x) * pixel_size];
+
+	for (int y = r.position.y; y < r.position.y + r.size.y; y++) {
+		for (int x = r.position.x; x < r.position.x + r.size.x; x++) {
+			uint8_t *dst = &dst_data_ptr[(y * width + x) * pixel_size];
+
+			for (int k = 0; k < pixel_size; k++) {
+				dst[k] = rect_first_pixel_ptr[k];
+			}
+		}
+	}
+}
```

Comment:
Was thinking about it, realized you could do something like this:
```suggestion
	// Fill in the first row, skip the first pixel (it's already been filled).
	for (int x = 1; x < r.size.x; x++) {
		memcpy(rect_first_pixel_ptr + x, rect_first_pixel_ptr, pixel_size);
	}
	
	// Fill in the remaining rows, skip the first row (it's already been filled).
	for (int y = 1; y < r.size.y; y++) {
		memcpy(rect_first_pixel_ptr + y * width, rect_first_pixel_ptr, width * pixel_size);
	}
```
I haven't tested it so I'd be amazed if it worked first time, but hopefully you get the idea.

A slightly more optimize version for filling in the first row (same optimization could be applied for the fill method, since the fill method fills a contiguous region of memory).
```suggestion
	// Fill in the first row, skip the first pixel (it's already been filled).
	for (int x = 1, stride = 1; x < r.size.x; x += stride) {
		if (x + stride > r.size.x) {
			memcpy(rect_first_pixel_ptr + x, rect_first_pixel_ptr, r.size.x - x);
			break;
		}
		memcpy(rect_first_pixel_ptr + x, rect_first_pixel_ptr, pixel_size * stride);
		// Copy twice as much each time to make as few `memcpy()` calls as possible.
		stride *= 2;
	}
	// Fill in the remaining rows, skip the first row (it's already been filled).
	for (int y = 1; y < r.size.y; y++) {
		memcpy(rect_first_pixel_ptr + y * width, rect_first_pixel_ptr, width * pixel_size);
	}
```
(At least, I assume it's more optimized, but I'm not sure by how much)